[PP] Fix zero bubble composability with DP

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134014
* __->__ #134052

Moved all the backward functions (`stage_backward_input`, `stage_backward_weight`, `stage_backward`) under the same `backward_maybe_with_nosync` function which controls the logic of the data parallel wrappers.

FSDP was not working with zero bubble PP because there will be twice as many "backward" calls and we update the weight gradients after `autograd.grad` is called. As a result, we need to manually call the FSDP `post_backward_hook()` after the weights have the correct gradients.

Fixes the tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`

cc @XilunWu @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
[PP] Fix zero bubble composability with DP

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`



cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`


cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

Currently hitting an error in FSDP

```
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/stage.py", line 500, in <lambda>
    return lambda: stage_backward_weight(
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/_backward.py", line 259, in stage_backward_weight
    raise RuntimeError(f"{grad_acc=} not in {all_dweights.keys()=}")
RuntimeError: grad_acc=<AccumulateGrad object at 0x7f7d63f42c20> not in all_dweights.keys()=dict_keys([<AccumulateGrad object at 0x7f7d63f0b730>, <AccumulateGrad object at 0x7f7d63f0b5e0>, <AccumulateGrad object at 0x7f7d63f0b790>, <AccumulateGrad object at 0x7f7d63f0b760>, <AccumulateGrad object at 0x7f7d63f0b7c0>, <AccumulateGrad object at 0x7f7d63f0b7f0>, <AccumulateGrad object at 0x7f7d63f0b820>, <AccumulateGrad object at 0x7f7d63f0b850>])
```

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

Currently hitting an error in FSDP

```
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/stage.py", line 500, in <lambda>
    return lambda: stage_backward_weight(
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/_backward.py", line 259, in stage_backward_weight
    raise RuntimeError(f"{grad_acc=} not in {all_dweights.keys()=}")
RuntimeError: grad_acc=<AccumulateGrad object at 0x7f7d63f42c20> not in all_dweights.keys()=dict_keys([<AccumulateGrad object at 0x7f7d63f0b730>, <AccumulateGrad object at 0x7f7d63f0b5e0>, <AccumulateGrad object at 0x7f7d63f0b790>, <AccumulateGrad object at 0x7f7d63f0b760>, <AccumulateGrad object at 0x7f7d63f0b7c0>, <AccumulateGrad object at 0x7f7d63f0b7f0>, <AccumulateGrad object at 0x7f7d63f0b820>, <AccumulateGrad object at 0x7f7d63f0b850>])
```

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

Currently hitting an error in FSDP

```
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/stage.py", line 500, in <lambda>
    return lambda: stage_backward_weight(
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/_backward.py", line 259, in stage_backward_weight
    raise RuntimeError(f"{grad_acc=} not in {all_dweights.keys()=}")
RuntimeError: grad_acc=<AccumulateGrad object at 0x7f7d63f42c20> not in all_dweights.keys()=dict_keys([<AccumulateGrad object at 0x7f7d63f0b730>, <AccumulateGrad object at 0x7f7d63f0b5e0>, <AccumulateGrad object at 0x7f7d63f0b790>, <AccumulateGrad object at 0x7f7d63f0b760>, <AccumulateGrad object at 0x7f7d63f0b7c0>, <AccumulateGrad object at 0x7f7d63f0b7f0>, <AccumulateGrad object at 0x7f7d63f0b820>, <AccumulateGrad object at 0x7f7d63f0b850>])
```

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -232,16 +235,19 @@ def build_stage(stage_idx, num_stages):
                     parts[0] = str(int(parts[0]) + offset)
                     name = ".".join(parts)
                     ref_p = ref_parameters[name]
+                    # TODO: WITH FSDP
```
p.grad remains plain `torch.Tensor` because last_forward=False, see repros in https://github.com/pytorch/pytorch/pull/134375
```
[rank0]:V0823 15:41:06.855000 4139345 torch/distributed/pipelining/stage.py:533] [0] backward_type='full' set_is_last_backward(last_backward=False) set_requires_gradient_sync(last_backward=False) self._seen_bwd_chunks=15, self.chunks=8


[rank3]:V0823 15:41:06.922000 4139348 torch/distributed/pipelining/stage.py:533] [1] backward_type='weight' set_is_last_backward(last_backward=True) set_requires_gradient_sync(last_backward=True) self._seen_bwd_chunks=15, self.chunks=8
```

weifengpy commented on torch/distributed/pipelining/stage.py:
```diff
@@ -468,14 +468,40 @@ def forward_maybe_with_nosync(self, *args, **kwargs):
         This helper should adapt any pipeline parallel schedule to work with common/supported data parallel libraries.
         """
-        last_backward = self._seen_bwd_chunks == self.chunks - 1  # type: ignore[operator]
+        if backward_type == "full":
```
ditto to above: for PP + FSDP2, I do see `backward_type == 'full'`. rank0 and rank2 has `last_backward=False` in last backward. so their parameters are not sharded

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

Currently hitting an error in FSDP

```
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/stage.py", line 500, in <lambda>
    return lambda: stage_backward_weight(
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/_backward.py", line 259, in stage_backward_weight
    raise RuntimeError(f"{grad_acc=} not in {all_dweights.keys()=}")
RuntimeError: grad_acc=<AccumulateGrad object at 0x7f7d63f42c20> not in all_dweights.keys()=dict_keys([<AccumulateGrad object at 0x7f7d63f0b730>, <AccumulateGrad object at 0x7f7d63f0b5e0>, <AccumulateGrad object at 0x7f7d63f0b790>, <AccumulateGrad object at 0x7f7d63f0b760>, <AccumulateGrad object at 0x7f7d63f0b7c0>, <AccumulateGrad object at 0x7f7d63f0b7f0>, <AccumulateGrad object at 0x7f7d63f0b820>, <AccumulateGrad object at 0x7f7d63f0b850>])
```

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

Currently hitting an error in FSDP

```
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/stage.py", line 500, in <lambda>
    return lambda: stage_backward_weight(
  File "/home/howardhuang/local/pytorch/torch/distributed/pipelining/_backward.py", line 259, in stage_backward_weight
    raise RuntimeError(f"{grad_acc=} not in {all_dweights.keys()=}")
RuntimeError: grad_acc=<AccumulateGrad object at 0x7f7d63f42c20> not in all_dweights.keys()=dict_keys([<AccumulateGrad object at 0x7f7d63f0b730>, <AccumulateGrad object at 0x7f7d63f0b5e0>, <AccumulateGrad object at 0x7f7d63f0b790>, <AccumulateGrad object at 0x7f7d63f0b760>, <AccumulateGrad object at 0x7f7d63f0b7c0>, <AccumulateGrad object at 0x7f7d63f0b7f0>, <AccumulateGrad object at 0x7f7d63f0b820>, <AccumulateGrad object at 0x7f7d63f0b850>])
```

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented on torch/distributed/pipelining/stage.py:
```diff
@@ -489,21 +515,41 @@ def backward_maybe_with_nosync(self, bwd_kwargs: Dict):
+                    fsdp_state = fully_shard.state(fsdp_module)
+                    for state in fsdp_state._state_ctx.all_states:
+                        if state._fsdp_param_group:
+                            state._fsdp_param_group.post_backward()
```
is this advocating a new fsdp api, like `model.reduce_grad` ?

H-Huang replied:
I guess this new api could be useful if users call `autograd.grad()` instead of `autograd.backward()` and still want to reduce the gradients like what I am doing.

In practice, I'm not sure how needed this is for FSDP users, maybe my case is just a pretty unconventional use case.

weifengpy replied:
I am asking `._state_ctx` and `._fsdp_param_group` are private. there is some risk of future change. but it's not a immeidate blocker for this PR

wconstab replied:
It's probably better if we use a private api and rely on ci composability tests to keep it working rather than make a new public api that we don't really need

kwen2501 replied:
Agree. Would be nice if we don't have to get into FSDP details.

H-Huang committed:
Update on "[PP] Fix zero bubble composability with DP"


Moved all the backward functions (`stage_backward_input`, `stage_backward_weight`, `stage_backward`) under the same `backward_maybe_with_nosync` function which controls the logic of the data parallel wrappers.

FSDP was not working with zero bubble PP because there will be twice as many "backward" calls and we update the weight gradients after `autograd.grad` is called. As a result, we need to manually call the FSDP `post_backward_hook()` after the weights have the correct gradients.

Fixes the tests:
`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_FSDP_ScheduleClass0_use_new_runtime_False`

`python test/distributed/_composable/test_composability/test_pp_composability.py ComposabilityTest.test_manual_with_data_parallel_dp_type_DDP_ScheduleClass0_use_new_runtime_False`

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/distributed/pipelining/stage.py:
```diff
@@ -468,14 +468,40 @@ def forward_maybe_with_nosync(self, *args, **kwargs):
+        if full_backward:
+            last_backward = self._seen_bwd_chunks == self.chunks - 1  # type: ignore[operator]
+        else:
+            # For backwards are split into weight and input, we will see twice as many bwd_chunks
+            last_backward = self._seen_bwd_chunks == 2 * self.chunks - 1  # type: ignore[operator]
```
nit: then it seems `_seen_bwd_chunks` should be renamed to `_seen_bwd_calls`? Minor though.

kwen2501 reviewed (approved):
LGTM

H-Huang commented:
@pytorchbot merge -i
